module/mfrc522_parser.c

```diff
@@ -0,0 +1,178 @@
+	  .cmd = MFRC522_CMD_GEN_RANDOM },
+	{ .input = "version",
+	  .parameter_amount = 0,
+	  .cmd = MFRC522_CMD_GET_VERSION },
+};
+
+/**
+ * Get the command associated with a command name
+ *
+ * @param token Command name found during parsing
+ *
+ * @return Return a pointer to a command declared in the commands array, or NULL if no
+ *         command matched the input
+ */
+static struct command *cmd_from_token(const char *token)
+{
+	size_t i;
+
+	for (i = 0; i < MFRC522_CMD_AMOUNT; i++)
+		if (!strcmp(token, commands[i].input))
+			return &commands[i];
+
+	return NULL;
+}
+
+/**
+ * Parse a command with multiple arguments
+ *
+ * @param input Remaining user input, mutable and allocated via kmalloc() in mfrc522_parse()
+ * @param cmd Command returned by cmd_from_token()
+ *
+ * @return A complete MFRC522 command with allocated memory for the extra data
+ */
+static struct mfrc522_command *parse_multi_arg(char *input, struct command *cmd)
+{
+	// We only enter this function if arguments have been given to the input
+	u8 parameter_amount = 0;
+	char *token;
+	char *extra_data;
+	u8 extra_data_len;
+	int ret;
+
+	while (input) {
+		token = strsep(&input, MFRC522_SEPARATOR);
+		parameter_amount++;
+
+		// The first parameter is the extra data
+		if (parameter_amount == 1)
+			extra_data = kstrdup(token, GFP_KERNEL);
+
+		// The second parameter is the length of the extr data
+		if (parameter_amount == 2) {
+			ret = kstrtou8(token, 10, &extra_data_len);
+			if (ret == -EINVAL) {
+				pr_err("[MFRC522] Invalid parameter for Data length: Expected number but got %s\n",
+				       token);
+				goto err;
+			}
+
+			if (ret == -ERANGE) {
+				pr_err("[MFRC522] Invalid parameter for Data length: Expected Unsigned Byte (0 - 255) but got %s\n",
+				       token);
+				goto err;
+			}
+		}
+	}
+
+	pr_info("[MFRC522] Found %d parameters for command %s\n",
+		parameter_amount, cmd->input);
+
+	if (parameter_amount != cmd->parameter_amount) {
+		pr_err("[MFRC522] Invalid command: %s: Expected %d arguments but got %d\n",
+		       cmd->input, cmd->parameter_amount, parameter_amount);
+		goto err;
+	}
+
+	return mfrc522_command_init(cmd->cmd, extra_data, extra_data_len);
+
+err:
+	kfree(extra_data);
+	return NULL;
+}
+
+struct mfrc522_command *mfrc522_parse(const char *input, size_t len)
+{
+	// Inputs are always organized according to the same format
+	//
+	// <cmd>[:<extra_data>:<extra_data_len>]
+	//
+	// Therefore, we will either get a simple command without any colons (in which case
+	// we will only be able to run strsep on the string ONCE) or a command with colons
+	// in which case it is required for the length of the extra data to be present.
+	// Also, since strsep modifies its input, we need to create a copy of the user's
+	// input.
+
+	// The input's length does not account for a terminating NULL, but `strsep` expects
+	// it. Allocate one more byte for the NULL terminator
+	char *input_mut = kmalloc(len + 1, GFP_KERNEL);
```

Comment:
Can't we just parse with `strnchr` and not need a NUL terminator ? (that way also not need a kmalloc)